### PR TITLE
libvirt: support proper networking in user session

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

With the help of ```qemu-bridge-helper``` it's possible for VMs run under ```qemu:///session``` to have proper network access and not resort to user mode networking.

In my case that means I can dump VirtualBox!

~~I am not sure if ```"QEMU_PR_HELPER=/run/current-system/sw/bin/qemu-pr-helper"``` is really the best way to pass in the path to the PR helper but since the qemu package to use is configurable, we need a global path.~~ We are using ```QEMU_PR_HELPER=/run/libvirt/nix-helpers/qemu-pr-helper``` based on @volth's feedback. 

Cc: @fpletz 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by volth** [comment]
> I am not sure if "QEMU_PR_HELPER=/run/current-system/sw/bin/qemu-pr-helper" is really the best way to pass in the path to the PR helper but since the qemu package to use is configurable, we need a global path.

May be ```/run/libvirt```, it has symlinks to files of "qemu used by libvirt" (to fight numerous problems (#27857) such as ```qemu``` upgrade)

```qemu``` might be not in ```environment.systemPackages``` and thus not in ```/run/current-system/sw```, or a different version of ```qemu``` might be in ```environment.systemPackages```

**Comment by peterhoeg** [comment]
Thanks @volth - how about this version?

File: nixos/modules/virtualisation/libvirtd.nix [hunk 1]
```diff
@@ -177,17 +185,23 @@ in {
...
+          ln -s --force ${pkgs.OVMF.fd}/FV/OVMF_VARS.fd /run/libvirt/nix-ovmf/
         ''}
       '';
 
       serviceConfig = {
         Type = "notify";
         KillMode = "process"; # when stopping, leave the VMs alone
         Restart = "no";
+        LogsDirectory = "libvirt/qemu";
+        RuntimeDirectory = [ "libvirt/dnsmasq" "libvirt/nix-ovmf" ];
```
Comment (volth) [inline nixos/modules/virtualisation/libvirtd.nix]: does it make directories in /run or /var/lib ?
(i am not sure how does dnsmasq work and which directories it needs, but it seems that there was ```/var/lib/libvirt/dnsmasq``` and ```/run/libvirt/nix-ovmf``` before and now "libvirt/dnsmasq" "libvirt/nix-ovmf" are closer to each other)

Comment (peterhoeg) [reply nixos/modules/virtualisation/libvirtd.nix]: You're right - nicely spotted.

**Comment by peterhoeg** [comment]
Any reservations @Infinisil ?

File: nixos/modules/virtualisation/libvirtd.nix [hunk 2]
```diff
@@ -23,6 +23,9 @@ let
     ''}
     ${cfg.qemuVerbatimConfig}
   '';
+  dir = "/var/lib/${dirName}";
+  dirName = "libvirt";
```
Comment (volth) [inline nixos/modules/virtualisation/libvirtd.nix]: Any reason to introduce this level of indirection?
Is it a preparation for running multiple ```libvirt``` services simulteniously, each with own ```dirName``` and ```dir``` ?

File: nixos/modules/virtualisation/libvirtd.nix [hunk 3]
```diff
@@ -113,14 +116,27 @@ in {
...
 
   ###### implementation
 
   config = mkIf cfg.enable {
 
-    environment.systemPackages = with pkgs; [ libvirt libressl.nc cfg.qemuPackage ];
+    environment = {
+      # this file is expected in /etc/qemu and not sysconfdir (/var/lib)
+      etc."qemu/bridge.conf".text = (lib.concatStringsSep "\n" (map (e:
+        "allow ${e}") cfg.allowedBridges));
```
Comment (Infinisil) [inline nixos/modules/virtualisation/libvirtd.nix]:
```nix
{
  etc."qemu/bridge.conf".text = lib.concatMapStringsSep "\n" (e: "allow ${e}") cfg.allowedBridges;
}
```

Comment (peterhoeg) [reply nixos/modules/virtualisation/libvirtd.nix]: Nice, fixed.

File: nixos/modules/virtualisation/libvirtd.nix (hunk 2 above)
Comment (peterhoeg) [reply nixos/modules/virtualisation/libvirtd.nix]: Not a good one, no. Fixed.

Any other concerns?

**Comment by volth** [comment]
> I am not sure if "QEMU_PR_HELPER=/run/current-system/sw/bin/qemu-pr-helper" is really the best way to pass 

Maybe ```/run/libvirt/bin/qemu-pr-helper``` will be better.
The concern here is: ```libvirtd``` does not restart on ```nixos-rebuild switch```, but "/run/current-system/sw/bin/qemu-pr-helper" might change if ```nixos-rebuild switch``` changes version of ```qemu```

File: nixos/modules/virtualisation/libvirtd.nix (hunk 2 above)
Comment (volth) [reply nixos/modules/virtualisation/libvirtd.nix]: I meant not only ```dir```, but also ```dirName = "libvirt";``` and questionable ```s/libvirt/${dirName}/g``` across the whole file.

Comment (peterhoeg) [reply nixos/modules/virtualisation/libvirtd.nix]: I'm sure I replied to this earlier. but clearly not. 

I'm always having a hard time guessing when it's ```libvirt``` and when it's ```libvirtd```, so I thought to simply make things easier for the next person by moving into a variable.

**Comment by peterhoeg** [comment]
> ah, it is already done in the code, I just answered to the obsolete description if the first post

Description updated. Any other concerns or are we good to go?

If you want the module fine-tuned, can we at least merge the change to derivation? I've been carrying this change on my side for a while and it causes a bunch of stuff to be rebuilt instead of just fetching from the cache.

**Comment by peterhoeg** [comment]
As I haven't heard any strong reservations, I'm taking the liberty of merging this.

**Pull request merged**